fdbserver/include/fdbserver/TenantEntryCache.actor.h

```diff
@@ -342,24 +342,78 @@ class TenantEntryCache : public ReferenceCounted<TenantEntryCache<T>>, NonCopyab
 
 	void put(const TenantNameEntryPair& pair) {
 		TenantEntryCachePayload<T> payload = createPayloadFunc(pair.first, pair.second);
-		if (mapByTenantId.find(pair.second.id) == mapByTenantId.end()) {
-			ASSERT(mapByTenantName.find(pair.first) == mapByTenantName.end());
+		auto idItr = mapByTenantId.find(pair.second.id);
+		auto nameItr = mapByTenantName.find(pair.first);
 
+		if (idItr == mapByTenantId.end() && nameItr == mapByTenantName.end()) {
+			// New tenant entry
 			mapByTenantId[pair.second.id] = payload;
 			mapByTenantName[pair.first] = payload;
 
 			TraceEvent(SevInfo, "TenantEntryCacheNewTenant", uid)
 			    .detail("TenantName", pair.first.contents().toString())
 			    .detail("TenantID", pair.second.id)
 			    .detail("TenantPrefix", pair.second.prefix);
+
+			CODE_PROBE(true, "TenantCache new entry");
+		} else if (idItr != mapByTenantId.end() && nameItr == mapByTenantName.end()) {
```

Comment:
This works, but I think you could write it more concisely like:

```
if (nameItr != mapByTenantName.end()) {
  mapByTenantId.erase(nameItr->value.entry.id);
}
if (idItr != mapByTenantId.end()) {
  mapByTenantName.erase(idItr->value.name);
}

mapByTenantId[pair.second.id] = payload;
mapByTenantName[pair.first] = payload;
```

You could also consolidate this down to a single trace event with details indicating what type of update this is, and you could write your code probes like:

```
CODE_PROBE(idItr == mapByTenantId.end() && nameItr == mapByTenantName.end(), "TenantCache new entry);
CODE_PROBE(idItr != mapByTenantId.end() && nameItr == mapByTenantName.end(), "TenantCache entry tenant name updated);
...
```

Reply:
The current implementation allows customized TraceEvent and would think would be useful to retain, I can avoid code-duplication from if that helps.

```
			mapByTenantId[pair.second.id] = payload;
			mapByTenantName[pair.first] = payload;
``` 

Will merge the CODE_PROBE as suggested.
Thanks

Reply:
I suspect there isn't much value in having a different trace event type for every permutation of options here, but maybe there is for distinguishing whether the entry is entirely new or not? Still, I think it would probably be sufficient if you wanted to capture that in a detail rather than its own type.

Reply:
I think either ways works fine, for the sake for making forward progress will avoid multiple TraceEvents, if needed we can always add it later (no change in functionality).